### PR TITLE
Fix EnC for compilations with dependencies using time-based versioning

### PR DESCRIPTION
**Pull request by tmat**
This reverts commit 7bb0b4ea56dd7cc08f89830bd049e96aaa7e7f89 and rebases the original change on top of stabilization.

All changes affecting version parsing have been committed before and are no longer part of this PR.

### CODE REVIEW

**Comment by tmat** [comment]
@dotnet-bot retest prtest/win/dbg/unit64 please
// Previous failure: http://dotnet-ci.cloudapp.net/job/roslyn_prtest_win_dbg_unit64/4670/
// Retest reason: https://github.com/dotnet/roslyn/issues/9408

**Comment by tmat** [comment]
@dotnet/roslyn-compiler Please review. 
@MattGertz Ask mode PR. I have validated end-to-end scenarios of editing projects with dependencies that have * in assembly version and EnC works as expected.

**Comment by MattGertz** [comment]
@tmat Per recent discussions this meets the bar, pending code review and tests passing.

**Comment by AlekseyTs** [comment]
LGTM

File: src/Compilers/CSharp/Portable/Symbols/AssemblySymbol.cs [hunk 1]
```diff
@@ -79,6 +79,16 @@ public override string Name
         public abstract AssemblyIdentity Identity { get; }
 
         /// <summary>
+        /// Assembly version pattern with wildcards represented by <see cref="ushort.MaxValue"/>,
+        /// or null if the version string specified in the <see cref="AssemblyVersionAttribute"/> doesn't contain a wildcard.
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/AssemblySymbol.cs]: Just curious: why null if there are no wildcards rather than the actual version?

Comment (tmat) [inline src/Compilers/CSharp/Portable/Symbols/AssemblySymbol.cs]: Could be either. @AlekseyTs preferred null.

File: src/Compilers/Core/Portable/MetadataReference/AssemblyIdentity.cs [hunk 2]
```diff
@@ -50,6 +50,25 @@ public sealed partial class AssemblyIdentity : IEquatable<AssemblyIdentity>
...
+            _publicKey = other._publicKey;
+            _lazyPublicKeyToken = other._lazyPublicKeyToken;
+            _isRetargetable = other._isRetargetable;
+
+            _version = version;
+            _lazyDisplayName = null;
+            _lazyHashCode = 0;
+        }
+
+        internal AssemblyIdentity WithVersion(Version version) => (version == _version) ? this : new AssemblyIdentity(this, version);
```
Comment (cston) [inline src/Compilers/Core/Portable/MetadataReference/AssemblyIdentity.cs]: Consider asserting `version != null`.

Comment (tmat) [inline src/Compilers/Core/Portable/MetadataReference/AssemblyIdentity.cs]: It is asserted in the constructor.

**Comment by cston** [comment]
LGTM

**Pull request merged**